### ISSUE TITLE
How to work with edge nested dependencies?

### ISSUE THREAD

**Issue by bcardiff** [crystal-lang/shards#132]
**TL;DR**  If A depends on B. B depends on C (using `~> x.y.z`). It seems that A is unable to pull edge unreleased version of C.

---

For a real example, if a project has the following exact dependency:

```
dependencies:
  mysql:
    github: crystal-lang/crystal-mysql
    version: 0.2.0
```

And mysql 0.2.0 shard.yml is:

```
dependencies: 
  db: 
    github: crystal-lang/crystal-db 
    version: ~> 0.2.0 
```

Is there a way to use unreleased features of crystal-db in the main project?
`crystal-db@master` has some small commits over 0.2.0. So 0.2.1 is not even released yet.

If the main project's shard.yml is changed to

```
dependencies:
  db:
    github: crystal-lang/crystal-db
    branch: master
  mysql:
    github: crystal-lang/crystal-mysql
    version: 0.2.0
```

Then crystal deps complains:

```
$ crystal deps
Updating https://github.com/crystal-lang/crystal-db.git
Updating https://github.com/crystal-lang/crystal-mysql.git
Error resolving db (*, ~> 0.2.0)
```

So shards is unable to combine the `branch: master` with `version: ~> 0.2.0` for crystal-db.

An alternative would be to have a crystal-mysql edge branch with branch: master dependency but that will require constant forced push to keep that branch over crystal-mysql master.

Maybe when resolving a single branch by name + version the resolution should be the branch name  **if** the version in shard.yml in the branch matches the version criteria.

Or maybe I just don't if this can be handled in another way.

**Comment by ysbaddaden** [comment]
Having to deal with git refs in addition to version numbers is a real headache. From a quick read, I'd expect the updated `shard.yml` to choose the `master` branch, as long as its version matches `~> 0.2.0`, yes.

**Comment by bcardiff** [comment]
Yes, that was what I was expecting. The shard.yml in db@master is untouched since the release of 0.2.0. I think the error is raised way before checking for db's shards.yml .

**Comment by RX14** [comment]
I too hit this problem, while waiting for https://github.com/crystal-lang/crystal-db/pull/40 to get merged, I try and use the shard.yml below, and it fails. The only workaround is to set `path: /data/programming/crystal-db/` instead of use a commit, as that seems to override everything.

```yml
dependencies:
  db:
    github: RX14/crystal-db
    commit: 0797ce72a3e237bd9b0782c154870413ed835d0c
  pg:
    github: will/crystal-pg
    version: ~> 0.13.1
```

This issue basically makes shards unusable for me, for this specific project (it's got two people working on it, using `path` isn't acceptable).

**Comment by ysbaddaden** [comment]
An it's broken message doesn't help; a reduced and reproducible test helps damn better. I spent some time to have a failing test. It's still not easy to fix, it requires that a Git refs has more weight than version numbers, but the former is a per-resolver constraint (which Shards doesn't have), when the latter is a cross-resolver constraint (which Shards has).

**Comment by ysbaddaden** [comment]
I can't reproduce in master.
